src/NuGet.Core/NuGet.Protocol.Core.v3/HttpSource/HttpSource.cs

```diff
@@ -568,6 +500,39 @@ public void Dispose()
             _httpClientLock.Dispose();
         }
 
+        private class ThrottledResponse : IDisposable
+        {
+            private IThrottle _throttle;
+            private readonly object _throttleLock = new object();
+
+            public ThrottledResponse(IThrottle throttle, HttpResponseMessage response)
+            {
+                if (response == null)
+                {
+                    throw new ArgumentNullException(nameof(response));
+                }
+
+                _throttle = throttle;
+                Response = response;
+            }
+            
+            public HttpResponseMessage Response { get; }
+
+            public void Dispose()
+            {
+                Response.Dispose();
+
+                lock (_throttleLock)
```

Comment:
why? `SemaphoreSlim.Release` is thread-safe, so should be its derivative `SemaphoreSlimThrottle`